File: Ejemplos/aula-java/src/main/java/es/joseluisgs/models/Curso.java
package es.joseluisgs.models;

import java.util.Arrays;

public class Curso {
    private static int contador = 0;
    private final int id = nextId();
    private final String nombre;
    private final String año;
    private final Tutor tutor;
    private final Alumno[] alumnos = new Alumno[30]; // es un array de 30 posicionesAlumno  null de momento

    public Curso(String nombre, String año, Tutor tutor) {
        this.nombre = nombre;
        this.año = año;
        this.tutor = tutor;
    }

    private static int nextId() {
        return ++contador;
    }

    private static void resetId() {
        contador = 0;
    }

    public int getId() {
        return id;
    }

    public String getNombre() {
        return nombre;
    }

    public String getAño() {
        return año;
    }

    public Tutor getTutor() {
        return tutor;
    }

    public Alumno[] getAlumnos() {
        return alumnos;
    }

//    public void setAlumnos(Alumno[] alumnos) {
//        this.alumnos = alumnos;
//    }


    @Override
    public String toString() {
        return "Curso{" +
                "id=" + id +
                ", nombre='" + nombre + '\'' +
                ", año='" + año + '\'' +
                ", tutor=" + tutor +
                ", alumnos=" + Arrays.toString(alumnos) +
                '}';
    }

    public void clear() {
        Arrays.fill(alumnos, null);
//        for (int i = 0; i < alumnos.length; i++) {
//            alumnos[i] = null;
//        }
        Alumno.resetId();
    }

    public int totalAlumnos() {
        int total = 0;
        for (Alumno alumno : alumnos) {
            if (alumno != null) {
                total++;
            }
        }
        return total;
    }

    private int getIndexById(int id) {
        for (int i = 0; i < alumnos.length; i++) {
            if (alumnos[i] != null && alumnos[i].getId() == id) {
                return i;
            }
        }
        return -1;
    }

    public Alumno createAlumno(Alumno alumno) {
        var indice = -1;
        for (int i = 0; i < alumnos.length; i++) {
            if (alumnos[i] == null) {
                indice = i;
                break;
            }
        }
        if (indice == -1) {
            return null;
        } else {
            alumnos[indice] = alumno;
            return alumno;
        }
    }

    public Alumno getAlumnoById(int id) {
        var indice = getIndexById(id);
        if (indice == -1) {
            return null;
        } else {
            return alumnos[indice];
        }
    }

    public Alumno updateAlumno(Alumno alumno) {
        var indice = getIndexById(alumno.getId());
        if (indice == -1) {
            return null;
        } else {
            alumnos[indice] = alumno;
            return alumno;
        }
    }

    public Alumno deleteAlumnoById(int id) {
        var indice = getIndexById(id);
        if (indice == -1) {
            return null;
        } else {
            var alumno = alumnos[indice];
            alumnos[indice] = null;
            return alumno;
        }
    }

    public Alumno deleteAlumno(Alumno alumno) {
        return deleteAlumnoById(alumno.getId());
    }

    public void imprimirListado() {
        for (Alumno alumno : alumnos) {
            if (alumno != null) {
                System.out.println(alumno);
            }
        }
    }

    public void getAllAlumnosOrderByNotasDesc() {
        // Método de ordenaciñon burbuja descedente por notas
        for (int i = 0; i < alumnos.length - 1; i++) {
            for (int j = 0; j < alumnos.length - 1 - i; j++) {
                // Cuidado con los nulos!!!!
                if (alumnos[j] != null && alumnos[j + 1] != null) {
                    if (alumnos[j].getCalificacion() < alumnos[j + 1].getCalificacion()) {
                        Alumno aux = alumnos[j];
                        alumnos[j] = alumnos[j + 1];
                        alumnos[j + 1] = aux;
                    }
                }
            }
        }

        imprimirListado();
    }

    private int getNumAlumnos() {
        int total = 0;
        for (Alumno alumno : alumnos) {
            if (alumno != null) {
                total++;
            }
        }
        return total;
    }

    private int getNumAprobados() {
        int total = 0;
        for (Alumno alumno : alumnos) {
            if (alumno != null && alumno.isAprobado()) {
                total++;
            }
        }
        return total;
    }

    public double getNotaMedia() {
        int total = 0;
        for (Alumno alumno : alumnos) {
            if (alumno != null) {
                total += alumno.getCalificacion();
            }
        }
        return (((total / getNumAlumnos()) * 100) / 100.0);
    }

    public String getInforme() {
        var numAlumnos = getNumAlumnos();
        var numAprobados = getNumAprobados();
        var numSuspensos = numAlumnos - numAprobados;
        var porcentageAprobados = (int) ((numAprobados * 100.0 / numAlumnos) * 100) / 100.0;
        var porcentageSuspensos = numSuspensos * 100.0 / numAlumnos;
        var notaMedia = getNotaMedia();

        String sb = "Curso: " + nombre +
                " Año: " + año + "\n" +
                " Tutor/a: " + tutor + "\n" +
                " Número de alumnos: " + numAlumnos + "\n" +
                " Número de aprobados: " + numAprobados + "\n" +
                " Número de suspensos: " + numSuspensos + "\n" +
                " Porcentaje de aprobados: " + porcentageAprobados + "\n" +
                " Porcentaje de suspensos: " + String.format("%.2f", porcentageSuspensos) + "\n" +
                " Nota media: " + notaMedia + "\n";
        return sb;

    }
}

File: Ejemplos/ProgramacionOrientadaObjetos/src/main/kotlin/aula/models/Curso.kt
package aula.models

data class Curso(val nombre: String, val año: String, val tutor: Tutor) {
    private val alumnado = Array<Alumno?>(30) { null }

    val id = nextId()

    companion object {
        private var contador = 0
        private fun nextId() = ++contador
        fun reset() {
            contador = 0
        }

    }

    override fun toString(): String {
        return "Curso(id='$id', nombre='$nombre', año='$año', tutor= '$tutor' alumnado=${alumnado.contentToString()})"
    }

    fun clear() {
        alumnado.fill(null)
        Alumno.reset()
    }

    fun totalAlumnos(): Int {
        var total = 0
        for (alumno in alumnado) {
            if (alumno != null) {
                total++
            }
        }
        return total
    }

    /**
     * Crea un alumno si hay espacio en el aula
     * @param alumno: Alumno a crear
     * @return Alumno creado o null si no se ha podido insertar
     */
    fun createAlumno(alumno: Alumno): Alumno? {
        // Buscamos una posición sin null
        var index = -1
        for (i in alumnado.indices) {
            if (alumnado[i] == null) {
                index = i
                break
            }
        }
        // Early return error
        return if (index == -1) {
            null
        } else {
            alumnado[index] = alumno
            alumno
        }
    }

    /**
     * Obtiene un alumno en base a su id
     * @param id del alumno
     * @return alumno encontrado o null si no existe
     */
    fun getAlumnoById(id: Int): Alumno? {
        // recorremos buscando el alumno
        /*for (alumno in aula) {
            if (alumno != null && alumno.id == id) {
                return alumno
            }
        }*/
        val index = getIndexById(id)
        return if (index != -1)
            alumnado[index]
        else
            return null
    }

    /**
     * Actualiza un alumno en base a su id
     * @param id del del alumno
     * @return alumno actualizado o null si no existe
     */
    fun updateAlumno(alumno: Alumno): Alumno? {
        val index = getIndexById(alumno.id)
        return if (index != -1) {
            alumnado[index] = alumno
            alumno
        } else
            null
    }

    /**
     * Elimina un alumno en base a su id
     * @param id del alumno
     * @return alumno eliminado o null si no existe
     */
    fun deleteAlumnoById(id: Int): Alumno? {
        val index = getIndexById(id)
        // Early return error
        return if (index != -1) {
            val alumno = alumnado[index]
            alumnado[index] = null
            alumno
        } else
            null
    }

    /**
     * Elimina un alumno en base a su id
     * @param id del alumno
     * @return alumno eliminado o null si no existe
     */
    fun deleteAlumno(alumno: Alumno): Alumno? {
        val index = getIndexById(alumno.id)
        // Early return error
        return if (index != -1) {
            alumnado[index] = null
            alumno
        } else
            null
    }


    /**
     * Obtiene el índdice del objeto con id
     * @param id id a buscar
     * @return indice si existe, -1 si no existe
     */
    private fun getIndexById(id: Int): Int {
        for (i in alumnado.indices) {
            if (alumnado[i]?.id == id)
                return i
        }
        return -1
    }


    /**
     * Devuelve todos los alumnos ordenados por nota de mayor a menor
     */
    fun getAllAlumnosOrderByNotasDesc() {
        // Ordenar por nota usando burbuja desc
        for (i in 0 until alumnado.size - 1) {
            for (j in 0 until alumnado.size - 1 - i) {
                /*if (aula[j]!=null && aula[j + 1]!=null && aula[j]!!.califcacion < aula[j + 1]!!.califcacion) {
                    val aux = aula[j]
                    aula[j] = aula[j + 1]
                    aula[j + 1] = aux
                }*/
                if ((alumnado[j]?.califcacion ?: -1.0) < (alumnado[j + 1]?.califcacion ?: -1.0)) {
                    val aux = alumnado[j]
                    alumnado[j] = alumnado[j + 1]
                    alumnado[j + 1] = aux
                }
            }
        }

        // Imprimir --> se puede mover fuera
        imprimirListado()
    }

    private fun imprimirListado() {
        for (alumno in alumnado) {
            if (alumno != null) {
                println(alumno)
            }
        }
    }

    fun getInforme(): String {
        val numAlumnos = getNumAlumnos()
        val numAprobados = getNumAprobados()
        val numSuspensos = numAlumnos - numAprobados
        val porcentageAprobados = numAprobados * 100.0 / numAlumnos
        val porcentageSuspensos = numSuspensos * 100.0 / numAlumnos
        val notaMedia = getNotaMedia()
        return """
        Curso: $nombre
        Año: $año
        Tutor/a: $tutor
        Número de alumnos: $numAlumnos
        Número de aprobados: $numAprobados
        Número de suspensos: $numSuspensos
        Porcentaje de aprobados: $porcentageAprobados%
        Porcentaje de suspensos: $porcentageSuspensos%
        Nota media: $notaMedia
    """.trimIndent()
    }

    fun getNotaMedia(): Double {
        var suma = 0.0
        var numAlumnos = 0
        for (alumno in alumnado) {
            if (alumno != null) {
                suma += alumno.califcacion
                numAlumnos++
            }
        }
        return (suma / numAlumnos * 100).toInt() / 100.0
    }

    fun getNumAprobados(): Int {
        var num = 0
        for (i in alumnado.indices) {
            if (alumnado[i]?.isAprobado == true) {
                num++
            }
        }
        return num
    }

    private fun getNumAlumnos(): Int {
        var num = 0
        for (i in alumnado.indices) {
            if (alumnado[i] != null) {
                num++
            }
        }
        return num
    }

}